ReadMe.txt

Send images using WWarp to a WHDLoad developer
----------------------------------------------

If you get requested to send images of your game using WWarp to a WHDLoad
install author, just do the following:
 - copy the WWarp executable to somewhere it is reachable via your PATH, e.g.
   copy it to C:
 - enter the directory where you like to store the images to create, there
   should be some MB of free store
 - insert the disk to image into your floppy drive and run WWarp with the 
   name of the image to create as argument, if there are multiple disks for
   the game add the disk number to the name, e.g. "wwarp robocop2-d1"
 - if you are using not DF0: but a other drive add the unit number on the
   argument list, e.g. "wwarp robocop2-d1 unit=1" for DF1:
 - each image created by wwarp will have a size between 1 and 4.5 MB
 - at the end pack all images into one lha archive and email this archive
   to the WHDLoad install author
Here an complete example of how to install WWarp and create images for the
game RoboCop2:
	cd <where you have unpacked the wwarp archive>
	copy wwarp c:
	cd <where you like to create your images>
	wwarp robocop2-d1
	wwarp robocop2-d2
	lha a robocop2 #?.wwp
This creates "robocop2.lha" which you can send by email.


Usage
-----

Synopsis:

	WWarp filename[.wwp] [command] [tracks] [args] [options...]

commands:

	C - create wwarp file (default)
	    this will read the disk in the floppy drive and create a image 
	    from, if [tracks] has been specified only these tracks will be 
	    read
	
	D - dump tracks
	    displays info about the specified track and displays the mfm-data
	    (or decoded data if type is not 'raw'), if there is a sync set in
	    the wwarp file the mfm-data will be displayed shifted to the first
	    sync
	    args = [sync[&mask]][,[syncno][,[len][,off]]]
	    	sync[&mask] - the track will be shiftet to that sync and then
	    		displayed
	    	syncno - if multiple syncs are found, the number of the sync
	    		to use for display (first sync is 1)
	    	len - amount of bytes to display
	    	off - offset to start display, can be negative e.g. to view
	    		bytes before the sync, a bit offset will be specified
	    		using $1000.7 for example
	    	leadings args can be ommited, e.g. use ,,16 to see first the
	    	16 bytes

	F - try to decode the specified [tracks] as one of the known formats
	    (less restrictive than the routine used in command C for some
	    custom formats)

	G - save contents as image file, all decoded tracks will be saved in
	    their size, track of type RAW or with a data length of 0 will be
	    saved as $1600 bytes with the contents 'TDIC' (as used with the DIC
	    program), you can save a part of the image or skip tracks by
	    selecting only the whished tracks
	
	I - print informations about wwarp file and tracks contained
	
	L - set track length
	    sets the data length which should be written back to disk. this 
	    length is stored in the wwarp file and will be used on command 'W'
	    (see below), this operation does not modify the stored track data
	    in the wwarp file, this command doesn't work with tracks which are
	    not of type 'raw' of course
	
	P - pack wwarp file
	    this removes all unneccessary data from the wwarp file, all tracks
	    of type 'raw' and a valid 'Length' and 'Sync' will be stored
	    starting with 'Sync' and the length 'Length', removed data is lost
	    forever, so be careful

	R - remove tracks from wwarp file
	    deletes the specified [tracks] from the wwarp file, the tracks are
	    lost forever, so be careful

	S - save tracks
	    each track will be saved to a separate file, the filename contains
	    the track number and the format, if the format is raw-single the
	    doubled length will be saved, if the format is raw and a sync is
	    set the data will be shifted to this sync, if length is also set only
	    length data bytes are saved
	
	W - write wwarp file to disk

	Y - set sync
	    sets the syncronization to write the track with sync back to a
	    floppy disk, the sync can consist of upto 16 byte data and a mask
	    of the same length. both, data and mask must be given in
	    hexadecimal notation: "data&mask", the mask can be omitted and
	    will then calculated from the data value (rounded up to byte
	    boundaries)
	    args = [sync[&mask]][,syncno]
	    	sync[&mask] - syncronization to set
	    	syncno - if multiple syncs are found, the number of the sync
	    		to use (first sync is 1)

	Z - write custom format or raw data from a data file to the disk
	    This command does not operate on a wwarp file!
	    The filename specifies the datafile which will be written to the
	    specified track (only one track must be set). The format to be
	    written must be specifed as number. The file length must match the
	    tracklength for the given format. The supported formats and their
	    number are displayed when wwarp is called without arguments.
	    To write raw-mfm data the format number 0 must be used. Then the
	    first 16 bytes are considered the sync and will be used for verify.
	    e.g. wwarp mynewtrack.47 z 47 17 (writes ZZKJA format)

tracks:
	will be used to specify which tracks should be affected by the command
	 1-5		tracks 1,2,3,4,5
	 2,90		tracks 2 and 90
	 2*2		tracks 2,4,...,156,158
	 10-20*5	tracks 10,15,20
	 1-5,7,99-104*2 tracks 1,2,3,4,5,7,99,101,103
	 *	 	all tracks

options:

	BPT=BytesPerTrack/K - amount of bytes to rawread and write to the
		wwarp file if nothing will be detected, default $6c00

	DBG/K/N - enable debugging messages (see chapter below for more infos)

	Force/S - enable detection of formats with flag force

	Import/K - import alien file via trackwarp.library, must be used with
		in conjunction with command C

	NoFmt/K - don't try to detect known formats which are specified,
		formats are given by their number as listed when WWarp is
		called without arguments, multiple formats are delimited
		with comma, e.g. NoFmt=16,30,33 disables detection of
		slackskin, twilight1 and twilight2 formats

	NoStd/S - don't try to detect any known formats

	NV=NoVerify/S - disables verify on write operations, verify is
		supported with all modes!

	RC=RetryCnt/N/K - number of read retries before a track is stored as
		'nonsingle' 'raw', default 6

	SYBIL/S - use SYBIL hardware to write long track formats and formats
		with variable bitcell densities (see chapter below for more
		infos)

	Unit/N/K - trackdisk.device unit number, use 1 for DF1: etc.


Tips & Tricks
-------------

* use 'wwarp' without arguments for a short decription of the available
  options and to display all supported custom formats

* to write back an image:
  for that are usually multiple steps necessary because you have to say WWarp
  how to write back the track data.
  first step:
    'wwarp filename f' - that command forces WWarp to try to decode all
    tracks, thats recommended because WWarp by default saves standard AmigaDos
    tracks in decoded format only if the mfm data is absolutely clean. most
    format programs are weak in their job, e.g. formatting not the full track
    to save some miliseconds or do not init the sector headers correctly.
    because such differences could be used as a copy protection WWarp saves
    such tracks as raw by default. anyway usually you can use the F(orce)
    command to ignore such specials.
  second step:
    'wwarp filename y 2-159 9521000044&ffff0000ff' - set the sync, that must
    be done for all tracks which are not detected as a supported format by
    WWarp. WWarp needs that information to know where the track data starts.
    You should make sure that the specified sync is only one time contained
    on the track, otherwise WWarp may start on the wrong sector. If the disk
    format uses multiple sectores you have to set the sync that it refers the
    first sector after the inter sector gap. For example if you would like to
    write back a standard AmigaDOS track (which is unnecessary because WWarp
    knows that format ;-) you had to use:
    448944895500000555000001&ffffffffff000055ff000055
  third step:
    'wwarp filename l 2-159 $3050' - set the write length, that step is optional
    but recommended, WWarp needs that information to know how many bytes must
    be written to the disk. when writing WWarp will check if your drive is
    capable of writing the requested amount of bytes, e.g. it will say you if
    you try to write a long track on a normal drive
    if you don't set the length WWarp will write as much bytes as possible
    depending on the drive speed
  forth step:
    'wwarp filename w' - that will write back the tracks to disk, if you have
    prepared the wwarp file correctly and there are no specials preventing the
    write back operation you will have another duplicate of your original and
    saved this piece of software history for your sons and daughters ;)

* to check if the right sync has been set use:
	wwarp filename d 2-159 ,,32
  which will display the first 32 mfm bytes of the selected tracks, you will
  also see if the sync has been found and if its present multiple times

* if a disk is very weak and you have problems to warp it increase the
  RetryCnt, for example use:
  	wwarp filename c 10-14*2 RetryCnt=50
  get some coffee, watch some tv and after that maybe wwarp could read the
  tracks after several retries

* when WWarp detects the Rob Northen 12 sector format it also calculates the
  lower 31 bit of the DiskKey (bit #31 cannot be calculated, but seems to be
  most time zero, in the warp file that bit is always set!), this key is saved
  in front of the decoded track data in the warp file, to see it just use the
  (I)nfo command
  wwarp removes the gap between the sectors when writing the format back, that
  makes it possible to write also the long track variants of the format on
  standard disk drives

Using High-Density Floppy+Disk
------------------------------

There is some support for using HD disks. There are several limitations but it
may be useful under some conditions.

* WWarp will automatically detect if there is an HD floppy in a HD drive and
  some operations will be performed slightly different

* the following applies to standard AmigaDOS tracks: a track on a HD is not
  $1600 but $2c00 bytes in length, that requires some work arounds, WWarp will
  still write $1600 for each track but with the offset as it would be a DD
  disk, that means that track #1 will be physically located on the second half
  of track #0, tracks in std format will not be written using TD_FORMAT but
  TD_WRITE, as result the tracks must be formatted correctly before

SYBIL
-----

The SYBIL hardware has been created by Jim Drew of Utilities Unlimited
International. The hardware was created to read/write 800KB Macintosh Disks
which uses different speed zones. It seems there was also a successor called
AMIA which could be used together with Emplant (Mac-Emulator) which was also
from UU. The hardware consists of two cards connected with some wires. One card
connects to the parallel port, the other to the video port. The SYBIL is
controlled via the parallel port and changes the clocking of the custom chips
via the genlock interface. Because the disk controller is part of these chips
also the bitcell timing on writing is affected by the genlock interface.
Using the SYBIL hardware with WWarp you can write tracks with standard drives
(300 rpm) from $3000 bytes upto $3580 bytes length.

Debug output
------------

Setting the option DBG/K/N you can increase the amount of messages wwarp will
print on several operations. Following the additional messages depending on
the value specified:

 1    * on write operation of custom formats check the mfm data for "11" and
	"0000" sequences which are illegal for mfm
      * on read print each unsuccessful tried custom format
 2    * on write operation of custom formats dump encoded mfm data
 3    *	on retrying raw read because track length could not be estimated the
	head will be moved, now instead the dot the cylinder will be printed

MORE QUESTIONS? so check the source ;-)

Full package of WWarp can obtained from http://whdload.de/whdload/
The source is accessible via: https://github.com/wepl/wwarp
Please report bugs via the GitHub page. Contributions are welcome too!



History.txt

***  1.29 (14.03.2020)
	- fix:	command Zap did not open trackdisk.device
	- new:	command Zap now also supports raw mfm data, first 16 bytes
		are considered as sync and used for verify
	- chg:	command Save on tracks of type RAW now honors a set sync, sync
		number and length and will shift and cut the mfm data to save
		accordingly
	- fix:	console handling improved, works also on vamos now
	source is accessible via: https://github.com/wepl/wwarp

***  1.28 (06.05.2008)
	- new:	supports 5.25" drives, 80 tracks, but not really tested (Mark)
	- chg:	tracklist will be compressed now, old WWarps cannot read these
		files if track 0 is not contained due a bug in old versions
	- chg:	builtin help improved a lot, more/less feature
	- chg:	new search routine for custom format detection which uses a
		lookup table, around 7 times faster as in version 1.27

***  1.27 (30.05.2007)
	- new:	option NoFmt/K added to exclude a list of formats from known
		formats detection
	- chg:	MFM2WWarp removed from distribution

***  1.26 (30.01.2006)
	- new:	format from the game SlackSkinAndFlint added (Codetapper)

***  1.25 (08.04.2005)
	- new:	support for trackwarp.library added, enables direct import of
		alien warp files
	- new:	option Force/S added to make WWarp to detect track formats
		with flag F during creating warp files using command C (e.g.
		while read from floppy or import), this obsolets aftwards
		usage of command F
	- chg:	packmodes do not work on specdata anymore, so packing is
		possible for tracks formats with special data

***  1.24 (16.11.2004)
	- new:	format rob northen copylock old, uses standard dos track
		with extra sync and data, decoder only, no write back
	- new:	MFM2WWarp supports also RNWarp and NOMADwarp

***  1.23 (28.10.2004)
	- fix:	limit support for hackdisk.device added, reading up to track
		161 works, 162/163 gives wrong results, writing crashes my
		machine, so I recommend to NOT use it (tested with 2.04)
	- fix:	could crash when not using asyncio and errors on file io

***  1.22 (04.10.2004)
	- new:	formats twilight2/3 added (Awesome/Psygnosis)

***  1.21 (06.07.2004)
	- fix:	write of non custom formats works again
	- fix:	SYBIL handling improved

***  1.20 (18.06.2004)
	- new:	MFM2WWarp supports CAPS/IPF images too (Psygore)
		MFM2WWarp uses asyncio.library if available

***  1.19 (28.04.2004)
	- fix:	degug code removed which caused illegal instruction exception
		on each custom format write back operation
	- note:	write back of RNCopyLock is still not working

***  1.18 (08.04.2004)
	- new:	command G to save contents of wwarp file as a dos disk image
	- new:	completed support for SYBIL hardware (calibration etc.)
	- new:	formats thalamus, goliath, beyond, hitec, mason (Codetapper)
	- new:	format Rob Northen Copylock, displays calculated ID's by
		several routines, write back only using SYBIL
	- chg:	MAXTRACKS increased to 168 which means upto cylinder 83 can
		be processed
	- chg:	simplified custom format interface
	- new:	support for MULTISYNC formats
	- new:	option DBG/K/N to set debug level, higher values causes wwarp
		to output debugging messages on several operations (see
		ReadMe.txt)
	- chg:	various minor changes and improvements

***  1.17 (02.12.2002)
	- new:	SpecialFX format added (Codetapper)
	- fix:	write back error introduced in 1.16

***  1.16 (02.12.2002)
	- chg:	again a bit faster scanning for custom formats
	- new:	custom formats Ocean, Twilight, Vision and ZZKJ added
		(Codetapper/Wepl)
	- new:	support for SYBIL hardware added, currently support is very
		limited, no speed adjustment, it will be improved later...
		many thanks to Galahad for the hardware :)
	- new:	command Z to write a data file as custom format to disk

***  1.15 (02.11.2002)
	- new:	program MFM2WWarp supports also old WWarp 1.3 file format
	- fix:	command Save did write wrong data on packed (LEQ/SLINC) tracks
		(Codetapper)
	- new:	pack mode SLEQ (sectors long equal)
	- new:	help screen lists also infos about all supported custom
		formats
	- chg:	faster scanning for custom formats

***  1.14 (28.08.2002)
	- chg:	program MFM2WWarp converts also files created by 'mfmwarp
		(Ferox), done by Psygore
	no changes on WWarp, only revision bumped

***  1.13 (09.05.2002)
	- new:	detected custom formats 'Turrican1/2/3', 'BloodMoney' and
		'Psygnosis1' added by Psygore
	- new:	program ADF2WWarp to convert UAE-MFM-ADF's to WWarp files
		by Psygore
	- fix:	command F(orce) clears the sync now

***  1.12 (15.12.2001)
	- new:	command I(nfo) prints all sync offsets
	- new:	detected custom formats 'PrimeMover', 'Beast1' and 'Beast2'
		added by Psygore

***  1.11 (06.11.2001)
	- fix:	bug in post gap calculation on raw write fixed (could possible
		create wrong extra syncs)
	- new:	command P(ack) implemented
	- new:	DOS drive inhibit added on write operation

***  1.10 (28.10.2001)
	- fix:	verify does also work with raw tracks which contain multiple
		sync's, it prints a message and will try all syncs for verify
	- new:	on multiple syncs the number of the sync to use can be
		specified on command Y and will saved in the wwarp file too
		and will be used on all other commands afterwards
	- chg:	the routine to decode standard dos tracks has been recoded to
		be more tolerant for format errors and variations, should at
		least detect all which is read by AmigaDOS

***  1.9 (20.09.2001)
	- new:	uses asyncio.library if available to speed up file operations
		(modifying an existing wwarp file is around 100% faster)
	- fix:	indexsync on read works now, all raw tracks will be written
		after index pulse now
	- new:	support for high density disk+drive added
	- new:	support for cylinder 80/81 added
	- chg:	misc other minor changes

***  1.8 (24.07.2001)
	- new:	Rob Northen 12 sector format directly supported
	- chg:	before reading trackdisk buffers will be flushed/cleared

***  1.7 (31.03.2001)
	- chg:	command dump understands bit offsets now, e.g. use

fmt_std.s
;*---------------------------------------------------------------------------
;  :Program.	fmt_std.s
;  :Contents.	decode standard dos track
;  :Author.	Wepl
;  :Version	$Id: fmt_std.s 1.9 2020/03/14 14:10:32 wepl Exp wepl $
;  :History.	11.11.02 separated from formats.s
;		12.11.02 optimized
;		20.02.04 new decode/encode parameters and rework
;		04.11.04 changes for rnclold support
;		12.01.07 wrong comments for odd/even exchanged
;  :Requires.	OS V37+, MC68020+
;  :Language.	68020 Assembler
;  :Translator. Barfly V2.9
;  :To Do.
;  :Info.
;---------------------------------------------------------------------------*
; $000 word	2	MFM value 0xAAAA AAAA (used for timing ?)
; $004 word	1	MFM value 0x4489
; $006 word	1	MFM value 0x4489
; $008 long	1	info (odd bits)
; $00c long	1	info (even bits)
;		decoded long is : ff TT SS SG
;			TT = track number ( 3 means cylinder 1, head 1)
;			SS = sector number ( 0 upto 10/21 ) sectors are not ordered !!!
;			SG = number of sectors before gap (including current one)
;		Example for cylinder 0, head 1 of a DD disk :
;			ff010009
;			ff010108
;			ff010207
;			ff010306
;			ff010405
;			ff010504
;			ff010603
;			ff010702
;			ff010801
;				-- inter-sector-gap here !
;			ff01090b (b means -1 ?)
;			ff010a0a (a means -2 ?)
; $010 long	4	sector label (odd)
; $020 long	4	sector label (even)
;			decoded value seems to be always 0
; $030 long	1	header checksum (odd)
; $034 long	1	header checksum (even)
;		computed on mfm longs between offsets 8 and $30, 2*(1+4) longs
; $038 long	1	data checksum (odd)
; $03c long	1	data checksum (even)
; $040 long	512	coded data (odd)
; $240 long	512	coded data (even)
; $440
;----------------------------------------
; decode standard amigados track
; IN:	D0 = ULONG length of raw mfm-data in bits
;	D1 = ULONG track number
;	D2 = ULONG offset in mfm buffer
;	A0 = APTR  mfm buffer (source)
;	A1 = APTR  decoded buffer (destination)
; OUT:	D0 = BOOL  true on success
;	all other regs may scratched

	NSTRUCTURE	locals_decodedos,0
		NWORD	lc1_trknum
		NALIGNLONG
		NLABEL	lc1_SIZEOF

_decode_std	link	LOC,#lc1_SIZEOF

		or.w	#$ff00,d1
		move.w	d1,(lc1_trknum,LOC)

		move.l	d0,d5			;D5 = mfm-length
		move.l	a0,a2			;A2 = mfm-buffer
		move.l	a1,a3			;A3 = dest-buffer

	;fast check for standard dos track
		move.l	#$55555555,d4		;D4 = mfm decode
	;check first word of info
		bfextu	(4,a0){d2:16},d0
		bfextu	(8,a0){d2:16},d7
		and.w	d4,d0
		and.w	d4,d7
		add.w	d0,d0
		or.w	d7,d0
		cmp.w	d0,d1
		bne	.no
	;check sector labels, all must be cleared
		moveq	#7,d7
.lblchk		bfextu	(12,a0,d7.l*4){d2:32},d0
		and.l	d4,d0
		dbne	d7,.lblchk
		bne	.no

	;check if track is the last before inter sector gap
		bfextu	($440,a0){d2:32},d0
		cmp.l	#$44894489,d0
		beq	.no

	;search first track after inter sector gap
		move.l	d2,d0
		addq.l	#1,d0			;offset
		move.l	d5,d1			;buffer length
		lea	(_sync_std),a1		;44894489...
		bsr	_searchsync
		tst.l	d0
		bmi	.no

	;d2 = last sector before gap
	;d0 = first sector after gap
		move.l	d0,d6			;D6 = first sector after gap

	;check that inter sector gap is set to all zeros
		move.l	a2,a0
		sub.l	d2,d0			;sub start of last sector before gap
		sub.l	#($440-4)*8+33,d0	;length of gap in bits to check
		bmi	.no
		move.l	#$aaaaaaaa,d3
		bra	.cmpin
.cmp		bfextu	(a0){d6:32},d1
		cmp.l	d1,d3
		bne	.no
.cmpin		subq.l	#4,a0
		sub.l	#32,d0
		bcc	.cmp
		bfextu	(a0){d6:32},d1
		neg.l	d0
		lsl.l	d0,d1
		lsl.l	d0,d3
		cmp.l	d1,d3
		bne	.no

	;decode track
		move.l	#%11111111111,d4	;D4 = sectors
		move.l	#$55555555,d5		;D5 = mfm decode
		moveq	#11-1,d7		;D7 = sector loop count
		move.l	a2,a0			;A0 = mfm buffer
		subq.l	#4,a0			;first $aaaaaaaa

.sector		bsr	_getlw
		bclr	#31,d0
		cmp.l	#$2aaaaaaa,d0
		bne	.no
		bsr	_getlw
		cmp.l	#$44894489,d0
		bne	.no

		moveq	#0,d3			;D3 = chksum

		bsr	_getlwd			;sector header

		subq.b	#1,d0			;num sectors before gap
		cmp.b	d7,d0
		bne	.no

		lsr.l	#8,d0			;D0.b = sector number 0..10
		cmp.b	#10,d0
		bhi	.no
		bclr	d0,d4
		beq	.no			;same sector again
		moveq	#0,d1
		move.b	d0,d1
		mulu	#$200,d1
		lea	(a3,d1.l),a1		;A1 = decode destination

		lsr.l	#8,d0			;D0.w = format + track number
		cmp.w	(lc1_trknum,LOC),d0
		bne	.no

	;the sector labels, 4 longs decoded
		moveq	#3,d2
.label		bfextu	(a0){d6:32},d0
		bfextu	(16,a0){d6:32},d1
		eor.l	d0,d3
		eor.l	d1,d3
		and.l	d5,d0
		bne	.no
		and.l	d5,d1
		bne	.no
		addq.l	#4,a0
		dbf	d2,.label
		add.w	#16,a0

		move.l	d3,d2
		and.l	d5,d2
		bsr	_getlwd			;header chksum
		cmp.l	d2,d0
		bne	.no

		bsr	_getlwd			;data chksum
		move.l	d0,d3

	;the sector data
		moveq	#$200/4-1,d2
.data		bfextu	(a0){d6:32},d0
		bfextu	($200,a0){d6:32},d1
		eor.l	d0,d3
		and.l	d5,d0
		eor.l	d1,d3
		and.l	d5,d1
		add.l	d0,d0
		addq.l	#4,a0
		or.l	d0,d1
		move.l	d1,(a1)+
		dbf	d2,.data
		add.w	#$200,a0

		and.l	d5,d3
		bne	.no

		dbf	d7,.sector

		moveq	#-1,d0
		bra	.quit

.no		moveq	#0,d0

.quit		unlk	LOC
		rts

;----------------------------------------
; force decode standard amigados track
; IN:	D0 = ULONG length of raw mfm-data in bits
;	D1 = ULONG track number
;	D2 = ULONG offset in mfm buffer
;	A0 = APTR  mfm buffer (source)
;	A1 = APTR  decoded buffer (destination)
; OUT:	D0 = BOOL  true on success
;	all other regs may scratched

_decode_stdf	bsr	_trkchk_rnclold		;check if there are extra data of rnclold
		bne	_decode_stdf_no
_decode_stdf_nochk

	;decode track
		move.l	#%11111111111,d4	;D4 = sectors (0 means already decoded)
		move.l	#$55555555,d5		;D5 = mfm decode mask
		move.l	d2,d6			;D6 = offset
		moveq	#11-1,d7		;D7 = sector loop count
		move.l	a0,a2			;A2 = mfm-buffer
		move.l	a1,a3			;A3 = dest-buffer
		move.l	d0,a6			;A6 = mfm length

		bra	.in

.sector

	;search sync
		move.l	d6,d0			;offset
		move.l	a6,d1			;buflen
		move.l	a2,a0			;buffer
		lea	(_sync_stdf),a1		;sync
		bsr	_searchsync
		move.l	d0,d6			;D6 = offset
		bmi	.no

	;check that remaining mfm data is sufficent
.in		move.l	a6,d0
		sub.l	d6,d0			;in bits
		cmp.l	#$43c*8,d0
		blo	.no

		move.l	a2,a0			;A0 = mfm buffer
		addq.l	#4,a0			;skip $44894489

		bsr	_getlwd			;sector header

		lsr.l	#8,d0			;sector number 0..10
		cmp.b	#10,d0
		bhi	.no
		bclr	d0,d4
		beq	.no			;same sector again
		moveq	#0,d1
		move.b	d0,d1
		mulu	#$200,d1
		lea	(a3,d1.l),a1 		;A1 = dest buffer

		add.w	#32+8,a0		;skip sector label + chksum

		bsr	_getlwd			;data chksum
		move.l	d0,d3

		moveq	#$200/4-1,d2
.data		bfextu	(a0){d6:32},d0
		bfextu	($200,a0){d6:32},d1
		eor.l	d0,d3
		and.l	d5,d0
		eor.l	d1,d3
		and.l	d5,d1
		add.l	d0,d0
		addq.l	#4,a0
		or.l	d0,d1
		move.l	d1,(a1)+
		dbf	d2,.data

		and.l	d5,d3
		bne	.no

		add.l	#$43c*8,d6
		dbf	d7,.sector

		moveq	#-1,d0
		rts
.no
_decode_stdf_no
		moveq	#0,d0
		rts

;----------------------------------------

_getlw		bfextu	(a0){d6:32},d0
		addq.l	#4,a0
		eor.l	d0,d3
		rts

;----------------------------------------

_getlwd		bsr	_getlw
		move.l	d0,d1
		bsr	_getlw
		and.l	d5,d1
		and.l	d5,d0
		add.l	d1,d1
		or.l	d1,d0
		rts
